src/lib/Microsoft.Health.Fhir.Ingest/Service/MeasurementEventNormalizationService.cs

```diff
@@ -92,33 +94,53 @@ private async Task StartConsumer(ISourceBlock<EventData> producer, IAsyncCollect
                         foreach (var measurement in _contentTemplate.GetMeasurements(token))
                         {
                             measurement.IngestionTimeUtc = evt.SystemProperties.EnqueuedTimeUtc;
-                            await collector.AddAsync(measurement).ConfigureAwait(false);
-
-                            _log.LogMetric(
-                                IomtMetrics.NormalizedEvent(),
-                                1);
+                            createdMeasurements.Add(measurement);
                         }
                     }
-                    catch (OperationCanceledException)
-                    {
-                        cts.Cancel();
-                        throw;
-                    }
-#pragma warning disable CA1031
                     catch (Exception ex)
                     {
                         if (await errorConsumer(ex, evt).ConfigureAwait(false))
                         {
                             exceptions.Add(ex);
                         }
                     }
-#pragma warning restore CA1031
+
+                    return createdMeasurements;
+                });
+
+            var asyncCollectorConsumer = new ActionBlock<IMeasurement[]>(
+                async measurements =>
+                {
+                    try
+                    {
+                        await collector.AddAsync(measurements, cts.Token).ConfigureAwait(false);
+                        _log.LogMetric(
+                            IomtMetrics.NormalizedEvent(),
+                            measurements.Length);
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        cts.Cancel();
+                        throw;
+                    }
+                    catch (Exception ex)
+                    {
+                        exceptions.Add(ex);
+                    }
                 },
                 new ExecutionDataflowBlockOptions { MaxDegreeOfParallelism = _maxParallelism, SingleProducerConstrained = true, CancellationToken = cts.Token });

```

Comment:
>_maxParallelism [](http://example.com/codeflow?start=77&length=15)

Do you know what max Parallelism you tested with?  I suspect if we do more batching of events parallelism will be counter productive.  If we are only batching measurements  projected from the same event keeping the parallelism should be fine but if my understanding isn't correct and we are batching a different way then we may need to dig into this more.

Reply:
I think that this is current set to 4. Its my understanding that this will only apply to the TransformBlock. So we can be processing 4 device events in parallel. The outputs of the Measurement should all fan-in to the same BatchBlock/Action block.

Reply:
but theMaxDegreeOfParallelism is being set on the consumer.  Based on the change is there any reason for this to be anything other than 1?  The only reason I can think we would want this to be >1 is if we submit a batch and before it has returned we have already collected enough events to send our next batch.  I guess that is possible.

Reply:
I ended up setting this to 1 from 3.